File: Month00_CSharp/Workers/Configuration.cs
using System;

namespace Workers;

internal readonly record struct Configuration
{
    private const string MessageCountArg = "--message-count";
    private const string ThreadCountArg  = "--thread-count";

    public int MessageCount { get; init; }
    public int ThreadCount  { get; init; }

    public bool IsValid => MessageCount > 0 && ThreadCount > 0;

    public Configuration(string[] args)
    {
        if (args is not { Length: >= 4 })
        {
            Console.Error.WriteLine(
                $"Invalid input. Pass required arguments: '{MessageCountArg}', '{ThreadCountArg}'.");
            return;
        }

        for (var i = 0; i < args.Length && i + 1 < args.Length; i += 2)
        {
            switch (args[i])
            {
                case MessageCountArg:
                    MessageCount = int.TryParse(args[i + 1], out var m) ? m : 0;
                    break;
                case ThreadCountArg:
                    ThreadCount = int.TryParse(args[i + 1], out var t) ? t : 0;
                    break;
                default:
                    continue;
            }
        }
    }
}

File: 03_FSharp/Workers/Configuration.fs
namespace Workers

open System

type Configuration(args: string[]) =

    static let messageCountArg: string = "--message-count"
    static let threadCountArg: string = "--thread-count"

    let messageCount: int option =
        args
        |> Array.windowed 2
        |> Array.tryPick (fun pair ->
            if pair[0] = messageCountArg then
                match Int32.TryParse(pair[1]) with
                | true, m when m > 0 -> Some m
                | _ -> None
            else
                None)

    let threadCount: int option =
        args
        |> Array.windowed 2
        |> Array.tryPick (fun pair ->
            if pair[0] = threadCountArg then
                match Int32.TryParse(pair[1]) with
                | true, t when t > 0 -> Some t
                | _ -> None
            else
                None)

    do
        if messageCount.IsNone || threadCount.IsNone then
            Console.Error.WriteLine($"Invalid input. Pass required arguments: '{messageCountArg}', '{threadCountArg}'.")

    member _.MessageCount: int =
        if messageCount.IsSome && messageCount.Value > 0 then
            messageCount.Value
        else
            0

    member _.ThreadCount: int =
        if threadCount.IsSome && threadCount.Value > 0 then
            threadCount.Value
        else
            0

    member this.IsValid: bool = this.MessageCount > 0 && this.ThreadCount > 0
